### ISSUE TITLE
Maui performance issues under Android

### ISSUE THREAD

**Issue by memsom** [Mapsui/Mapsui#2597]
**Mapsui Version**
4.1.3, but also tried 4.1.5

**Mapsui Platform**
Maui Android (broken)
Maui UWP (seems to work)
Xamarin Forms (works)

**Device**
The test device is a Samsung Galaxy s21 running Android 14. Model is SM-G991B

**Describe the bug**
We have a test app that uses our API wrapper around Mapsui to present the same interface to all the platforms we support. This works on WPF, Xamarin Forms and now Maui. The code is essentially the same on all platforms, we just ship Assemblies that target the specific Frameworks.

The Maui port went well. Very little code changed. I have had it previously running and it seemed fine under .Net 7, but the new port if targeting .Net 8. We are seeing significant issues with performance under Android.

I was hopeful that I might be able to turn on GUI rendering, but setting the MapControl to have `UseGpu = true;` crashes the app.

**To Reproduce**

It would be extremely difficult for me to describe how our code works. But we essentially have multiple layers that render multiple Map objects on to the map. The demo app however just has a single map layer that is using OpenStreetMap.

The demo apps are using the same APIs and the same controls (it just has a different platform implementation under the hood that routes in the correct platform.)

**Expected behavior**

We would expect basic functionality, like moving/panning the map to be  smooth. However it is unusable under Android with Maui.

**Screenshots**

This movie probably shows the issue fairly well. I open the Xamarin Forms app then the Maui one. You can clearly see the difference.

https://github.com/Mapsui/Mapsui/assets/429330/9c067923-57d9-420b-b8ff-fcd270aa0970

**Additional context**

**Comment by memsom** [comment]
Part of the issue might be that we seem to not be able to use the accelerated GL rendering? I think SkiaSharp needs to be updated to 3.0.0-preview-3.1 to get that to work in Maui.

**Comment by pauldendulk** [comment]
I just ran the Mapsui.Samples.Maui from the debugger on a local device connected through usb and it performs well. Could you test how that performs on your device?

**Comment by pauldendulk** [comment]
What is remarkable in your movie (thanks, that helps a lot!) is that the map tiles are not updated, or much later. Perhaps the problem is with the tile fetch logic or the connection. Could you attach the logger, perhaps we can see what is going on https://mapsui.com/documentation/logging.html.

**Comment by memsom** [comment]
> Could you attach the logger, perhaps we can see what is going on https://mapsui.com/documentation/logging.html.

I will do that this morning. 

The same code works on multiple platforms, and so Android is the only one so far that shows this specific issue. I know we set up a number of layers and I don't really understand why Windows running the same Maui app code is "okay", but under Android it runs terribly. It doesn't seem to make any sense to me. I tried removing events we add in to track touches etc, but that didn't make much difference. I'm wondering if it might be to do with the animations bug? This is the same code that triggered that issue under Windows using WPF. I still think the non-CPU rendering might be causing it.

**Comment by charlenni** [comment]
Did you test the sample apps with your device? Do you see the same behavior? 

CPU is slower than GPU but in this case (only draw bitmaps) not so much.

**Comment by memsom** [comment]
Debug wise, I'm not hitting any errors. I am now removing all the code that adds any events or extra functionality to see if any of it is causing the slow downs.

**Comment by memsom** [comment]
> Did you try the samples? Did you see there the same behavior.

This is the next test. We are using .Net 8. Is that a potential problem? The Xamarin Forms app is obviously using Mono. It looks like debug builds do not use AOT, so wondering if that might be an issue also.

**Comment by memsom** [comment]
One other thing / the latest Skia Sharp preview has a working SKGLView under Android. How easy would it be to make that work with Mapsui? We are really going to need the acceleration at some point soon. Unfortunately it might be a show stopper as we have certain elements that are rendered to GL contexts.

**Comment by inforithmics** [comment]
@memson is the data fetched with the mapsui infrastructure or your own. 
The freezing user interface could be caused by code running on the UIThread. Android doesn’t like it when iOS code is run on the UIThread. A method to force a work item to the background thread is await …. ConfigureAwait(false)

**Comment by memsom** [comment]
> @memson is the data fetched with the mapsui infrastructure or your own. 

OSM. The same code is used on multiple platforms and the Maui version serums well under Windows. I might try it on iOS over the weekend - we don't currently support iOS. 

> The freezing user interface could be caused by code running on the UIThread. Android doesn’t like it when code is run on the UIThread.

It's possible. But the same code under Xamarin Forms shows no slow down. And I mean - same code. I basically made a Maui version of the Maui specific part of the API and cut and pasted the code from the Xamarin Forms version in to the Maui template all writhe changes to make the namespaces correct for Maui. If there is suck a slow down unfortunately it is something that is different between Xamarin Forms and Maui.

**Comment by inforithmics** [comment]
Can you debug the Android App and give the Android Output Log, because there UI Freezes are logged.

**Comment by memsom** [comment]
> Did you test the sample apps with your device? Do you see the same behavior?
> 
> CPU is slower than GPU but in this case (only draw bitmaps) not so much.

I am having a terrible time getting them to compile as my Visual Studio seems to only want to target .Net 8.0 at the moment.

**Comment by inforithmics** [comment]
For testing if GPU Rendering would improve performance you can use MapControl.UseGPU = true (For Android in Windows this won't work because for that SkiaSharp 3.0.0 is needed).

**Comment by memsom** [comment]
> For testing if GPU Rendering would improve performance you can use MapControl.UseGPU = true (For Android in Windows this won't work because for that SkiaSharp 3.0.0 is needed).

The only platforms we really care about are Windows and Android, so not having either is not ideal. I know having GPU rendering improves performance because we (I am guessing) already use it for Android in the Xamarin Forms version of the code we are currently migrating. There are no issues in the Xamarin Forms version of this code under Android, but without accelerated graphics I guess it is hard to know if that is the issue or not. Can I turn off the GPU rendering in Xamarin Forms also?

**Comment by charlenni** [comment]
Yes, with the same command, but setting the property UseGPU to false.

**Comment by charlenni** [comment]
And you have to set it before you create the MapControl.

**Comment by memsom** [comment]
> And you have to set it before you create the MapControl.

Cool. I will try in Monday. Our VPN seems to be down so I will need to wait for the office. I've just needed to reinstall Visual Studio 2022 as I ran out of disk space last week and uninstalling VS2019 hosed Maui from compiling properly. Unfortunately all nugets come from our internal cache, so I can't build our code now without the VPN. 

The main difference I can see is that we are building against .Net 8, and Mapsui seems to be .Net 7. I will try making our code target 7, and see if that helps.

**Comment by memson** [comment]
Hi Team,

You have cited my GitHub handle( @memson ) by mistake. I’m sure that you’re after Matt Emson and not me!! 😉

Best regards,
Michelle

From: Thomas Stocker ***@***.***>
Date: Friday, April 19, 2024 at 2:25 PM
To: Mapsui/Mapsui ***@***.***>
Cc: Michelle Emson ***@***.***>, Mention ***@***.***>
Subject: Re: [Mapsui/Mapsui] Maui performance issues under Android (Issue #2597)

@memson<https://github.com/memson> is the data fetched with the mapsui infrastructure or your own.
The freezing user interface could be caused by code running on the UIThread. Android doesn’t like it when iOS code is run on the UIThread. A method to force a work item to the background thread is await …. ConfigureAwait(false)
